In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.metrics import accuracy_score

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('Resources/lending_data.csv')

# Review the DataFrame
df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [3]:
df.columns

Index(['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt', 'loan_status'],
      dtype='object')

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.copy()
X.drop('loan_status', axis= 1, inplace =True)
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [5]:
# Review the y variable Series
len(y), y.head()

(77536,
 0        0
 1        0
 2        0
 3        0
 4        0
         ..
 77531    1
 77532    1
 77533    1
 77534    1
 77535    1
 Name: loan_status, Length: 77536, dtype: int64)

In [6]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [7]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [8]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify = y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [9]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', max_iter=200, random_state=1)


# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [10]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.9914878250103177
Testing Data Score: 0.9924164259182832


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [11]:
# Print the balanced_accuracy score of the model
#https://scikit-learn.org/stable/modules/generated/sklearn.metrics.balanced_accuracy_score.html#sklearn.metrics.balanced_accuracy_score
bal_acc= balanced_accuracy_score(y_test, predictions)
bal_acc

0.9442676901753825

In [12]:
# Generate a confusion matrix for the model
cm = confusion_matrix(y_test, predictions)
cm_df = pd.DataFrame(cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])
acc_score = accuracy_score(y_test, predictions)

In [13]:
# Print the classification report for the model
print("Confusion Matrix")
display(cm_df)
print(f"Accuracy Score : {acc_score}")
print(f"Ballanced Accuracy Score : {bal_acc}")
print("Classification Report")
print(classification_report(y_test, predictions))

Confusion Matrix


Predicted 0  Predicted 1
Actual 0        18679           80
Actual 1           67          558

Accuracy Score : 0.9924164259182832
Ballanced Accuracy Score : 0.9442676901753825
Classification Report
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The difference between the accuracy score and the balanced accuracy score (which starts with scoring the accuracy for each category and then averages these [1]) suggests that the predictions in one category are less reliable than in another. 

For the healthy loans, the f1-score, the precision, and the recall are very good, meaning that the model does a good job identifying healthy loans. However, the scores for the high-risk loans are not so good. This could be due to the training data having fewer samples for the model to learn from. The lowest score for the high-risk loans is the precision, which means that the model is missing some of {The f1-score is the harmonic mean of precision (the ratio of true positives over all positives [3]) and recall (which is the ratio of true positive over the Toal number of positives in sample (things that are positive and were categorized correctly, or incorrectly)[3]). } 

The value counts that we did earlier show that the data is about a 75%/25% split between healthy loans and high-risk loans. 

We can see that having a very large number of healthy loans (18759 of 75036 healthy loans in the data (values counts vs support numbers) has made the model extremely precises for healthy loans. Calculating the ratio of samples in the data to samples used to test we find that 25% of both classes were used. 

References: 
1. https://stackoverflow.com/questions/55548675/difference-between-balanced-accuracy-score-and-accuracy-score
2. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html
3. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html

In [14]:
# needed to insall the package for RandomOverSampler
#https://stackoverflow.com/questions/40008015/problems-importing-imblearn-python-package-on-ipython-notebook
#!pip install -U imbalanced-learn

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
over_samp = RandomOverSampler(sampling_strategy='minority', random_state= 1) 
#https://imbalanced-learn.org/dev/references/generated/imblearn.over_sampling.RandomOverSampler.html


# Fit the original training data to the random_oversampler model
X_res, y_res = over_samp.fit_resample(X, y)
# refrance the imbalanced learn, randomoversampler document page

In [16]:
# Count the distinct values of the resampled labels data
y_res.value_counts()

0    75036
1    75036
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [17]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
fier = LogisticRegression(solver='lbfgs', random_state=1)

# Fit the model using the resampled training data
fier.fit(X_res, y_res)

# Make a prediction using the testing data
predictions = fier.predict(X_res)
pred = pd.DataFrame({"Prediction": predictions, "Actual": y_res})
pred.head()

Prediction  Actual
0                0       0
1                0       0
2                0       0
3                0       0
4                0       0
...            ...     ...
150067           1       1
150068           1       1
150069           1       1
150070           1       1
150071           1       1

[150072 rows x 2 columns]

In [18]:
# Score the model
print(f"Training Data Score: {fier.score(X_res, y_res)}")
print(f"Testing Data Score: {fier.score(X_test, y_test)}")

Training Data Score: 0.9945026387334079
Testing Data Score: 0.9952022286421791


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
bal_acc2= balanced_accuracy_score(y_res, predictions)
bal_acc2

0.9945026387334079

In [20]:
# Generate a confusion matrix for the model
matrix = confusion_matrix(y_res, predictions)
matrix_df = pd.DataFrame(matrix, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"])
score = accuracy_score(y_res, predictions)

In [21]:
# Print the classification report for the model
print("Confusion Matrix")
display(matrix_df)
print(f"Accuracy Score : {score}")
print(f"Ballanced Accuracy Score : {bal_acc2}")
print("Classification Report")
print(classification_report(y_res, predictions))

Confusion Matrix


Predicted 0  Predicted 1
Actual 0        74614          422
Actual 1          403        74633

Accuracy Score : 0.9945026387334079
Ballanced Accuracy Score : 0.9945026387334079
Classification Report
              precision    recall  f1-score   support

           0       0.99      0.99      0.99     75036
           1       0.99      0.99      0.99     75036

    accuracy                           0.99    150072
   macro avg       0.99      0.99      0.99    150072
weighted avg       0.99      0.99      0.99    150072



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Aanswer: ** 
The random oversample method has increased the number of 'high-risk loans' which the model was tested on (75036 is more than the 2500 that was the total in the data set...). This is shown in a balanced accuracy score that has increased from 0.944 to 0.994. The new model also has increases in other categories. 

  

For the 'healthy loan' category the has been a slight decrease in the precision, recall and f1-scores. On the other hand, the scores for the 'high-risk loans' have increased by 12, 10, and 11 percentage points respectively. 

  

The weighted average has remained the same (being *weighted* it favored the bigger category in the previous model). The micro average is apparently better for imbalanced data sets as the calculation is of the *true* positives, *false* positive, and *false* negatives for *each label* in the data [1, 2].  

  

The accuracy for both models is 99% so they did a fair job of correct categorization (however if there was more test data for the 'high-risk loans' this would have likely brought the accuracy score down, a lot). 

  

  

References: 

1. https://stackoverflow.com/questions/55740220/macro-vs-micro-vs-weighted-vs-samples-f1-score
2. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html